### PR TITLE
Implement PSObject's IConvertible

### LINKED ISSUE #17627: Regression bug 7.2.0 - Unable to cast DateTime into DataRow DateTime column
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

```pwsh
$DataTable = [Data.DataTable]::new()
$null = $DataTable.Columns.Add('Date', [DateTime])
$Date = Get-Date
$Row = $DataTable.Rows.Add($Date)

$Row.Date = $Date # Error
$Row['Date'] = $Date # Error
$Row.Item('Date') = $Date # Works
$Row.Date = [DateTime]$Date # Works
```

Last working versions "PowerShell-7.2.0-preview.1" and "PowerShell-7.1.7".
First impacted version "PowerShell-7.2.0-preview.2".

### Expected behavior

`$Row.Date = $Date` will put the DateTime object in the DateTime column.

### Actual behavior

SetValueInvocationException: Exception setting "Date": "Unable to cast object of type 'System.Management.Automation.PSObject' to type 'System.IConvertible'.Couldn't store <04/07/2022 19:42:04> in Date Column.  Expected type is DateTime."

### Error details

```console
Exception             :
    Type           : System.Management.Automation.SetValueInvocationException
    ErrorRecord    :
        Exception             :
            Type    : System.Management.Automation.ParentContainsErrorRecordException
            Message : Exception setting "Date": "Unable to cast object of type 'System.Management.Automation.PSObject' to type 'System.IConvertible'.Couldn't store <04/07/2022
19:58:39> in Date Column.  Expected type is DateTime."
            HResult : -2146233087
        CategoryInfo          : NotSpecified: (:) [], ParentContainsErrorRecordException
        FullyQualifiedErrorId : CatchFromBaseAdapterSetValue
        InvocationInfo        :
            ScriptLineNumber : 5
            OffsetInLine     : 1
            HistoryId        : -1
            ScriptName       : C:\Users\illym\Desktop\Temp\PBug.ps1
            Line             : $Row.Date = $Date
            PositionMessage  : At C:\Users\illym\Desktop\Temp\PBug.ps1:5 char:1
                               + $Row.Date = $Date
                               + ~~~~~~~~~~~~~~~~~
            PSScriptRoot     : C:\Users\illym\Desktop\Temp
            PSCommandPath    : C:\Users\illym\Desktop\Temp\PBug.ps1
            CommandOrigin    : Internal
        ScriptStackTrace      : at <ScriptBlock>, C:\Users\illym\Desktop\Temp\PBug.ps1: line 5
                                at <ScriptBlock>, <No file>: line 1
    TargetSite     :
        Name          : BasePropertySet
        DeclaringType : System.Management.Automation.Adapter, System.Management.Automation, Version=7.3.0.5, Culture=neutral, PublicKeyToken=31bf3856ad364e35
        MemberType    : Method
        Module        : System.Management.Automation.dll
    Message        : Exception setting "Date": "Unable to cast object of type 'System.Management.Automation.PSObject' to type 'System.IConvertible'.Couldn't store <04/07/2022
19:58:39> in Date Column.  Expected type is DateTime."
    Data           : System.Collections.ListDictionaryInternal
    InnerException :
        Type           : System.ArgumentException
        Message        : Unable to cast object of type 'System.Management.Automation.PSObject' to type 'System.IConvertible'.Couldn't store <04/07/2022 19:58:39> in Date Column.
Expected type is DateTime.
        TargetSite     :
            Name          : set_Item
            DeclaringType : System.Data.DataColumn
            MemberType    : Method
            Module        : System.Data.Common.dll
        InnerException :
            Type       : System.InvalidCastException
            TargetSite :
                Name          : ChkCast_Helper
                DeclaringType : System.Runtime.CompilerServices.CastHelpers, System.Private.CoreLib, Version=7.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e
                MemberType    : Method
                Module        : System.Private.CoreLib.dll
            Message    : Unable to cast object of type 'System.Management.Automation.PSObject' to type 'System.IConvertible'.
            Source     : System.Private.CoreLib
            HResult    : -2147467262
            StackTrace :
   at System.Data.Common.DateTimeStorage.Set(Int32 record, Object value)
   at System.Data.DataColumn.set_Item(Int32 record, Object value)
        Source         : System.Data.Common
        HResult        : -2147024809
        StackTrace     :
   at System.Data.DataColumn.set_Item(Int32 record, Object value)
   at System.Data.DataRow.set_Item(DataColumn column, Object value)
   at System.Management.Automation.DataRowAdapter.PropertySet(PSProperty property, Object setValue, Boolean convertIfPossible)
   at System.Management.Automation.Adapter.BasePropertySet(PSProperty property, Object setValue, Boolean convert)
    Source         : System.Management.Automation
    HResult        : -2146233087
    StackTrace     :
   at System.Management.Automation.Adapter.BasePropertySet(PSProperty property, Object setValue, Boolean convert)
   at System.Management.Automation.PSProperty.SetAdaptedValue(Object setValue, Boolean shouldConvert)
   at System.Management.Automation.PSProperty.set_Value(Object value)
   at System.Management.Automation.Language.PSSetMemberBinder.SetAdaptedValue(Object obj, String member, Object value)
   at System.Dynamic.UpdateDelegates.UpdateAndExecute2[T0,T1,TRet](CallSite site, T0 arg0, T1 arg1)
   at System.Management.Automation.Interpreter.DynamicInstruction`3.Run(InterpretedFrame frame)
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
CategoryInfo          : NotSpecified: (:) [], SetValueInvocationException
FullyQualifiedErrorId : CatchFromBaseAdapterSetValue
InvocationInfo        :
    ScriptLineNumber : 5
    OffsetInLine     : 1
    HistoryId        : -1
    ScriptName       : C:\Users\illym\Desktop\Temp\PBug.ps1
    Line             : $Row.Date = $Date
    PositionMessage  : At C:\Users\illym\Desktop\Temp\PBug.ps1:5 char:1
                       + $Row.Date = $Date
                       + ~~~~~~~~~~~~~~~~~
    PSScriptRoot     : C:\Users\illym\Desktop\Temp
    PSCommandPath    : C:\Users\illym\Desktop\Temp\PBug.ps1
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, C:\Users\illym\Desktop\Temp\PBug.ps1: line 5
                        at <ScriptBlock>, <No file>: line 1
```

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Fix #17627
Replace #19071
Pester tests come from #19071

Implement IConvertible interface for PSObject.

Since PSObject can wrap an object of any type the implementation is dynamic too: it calls appropriate methods from PSObject's base class.

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by iSazonov** [comment]
@daxian-dbw Could you please help? I cannot find where fix PSInvokeMemberBinder.
Follow test fails:
```powershell
$semver = [System.Management.Automation.SemanticVersion]::new(1, 0, 0)
$ser = [System.Management.Automation.PSSerializer]::Serialize($semver)
$des = [System.Management.Automation.PSSerializer]::Deserialize($ser)

$des.<Ctrl-Space> # Shows all IConvertible methods including ToString() but next fails

$des.ToString()

InvalidOperation: Method invocation failed because [Deserialized.System.Management.Automation.SemanticVersion] does not contain a method named 'ToString'.
```

```
Get-Error

Exception             :
    Type        : System.Management.Automation.RuntimeException
    ErrorRecord :
        Exception             :
            Type    : System.Management.Automation.ParentContainsErrorRecordException
            Message : Method invocation failed because [Deserialized.System.Management.Automation.SemanticVersion] does not contain a method named 'ToString'.
            HResult : -2146233087
        TargetObject          : ToString
        CategoryInfo          : InvalidOperation: (ToString:String) [], ParentContainsErrorRecordException
        FullyQualifiedErrorId : MethodNotFound
        InvocationInfo        :
            ScriptLineNumber : 1
            OffsetInLine     : 1
            HistoryId        : 4
            Line             : $des.ToString()
            PositionMessage  : At line:1 char:1
                               + $des.ToString()
                               + ~~~~~~~~~~~~~~~
            CommandOrigin    : Internal
        ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
    TargetSite  :
        Name          : InvokeAdaptedMember
        DeclaringType : System.Management.Automation.Language.PSInvokeMemberBinder, System.Management.Automation, Version=7.3.0.0, Culture=neutral, PublicKeyToken=31bf
3856ad364e35
        MemberType    : Method
        Module        : System.Management.Automation.dll
    Message     : Method invocation failed because [Deserialized.System.Management.Automation.SemanticVersion] does not contain a method named 'ToString'.
    Data        : System.Collections.ListDictionaryInternal
    Source      : System.Management.Automation
    HResult     : -2146233087
    StackTrace  :
   at System.Management.Automation.Language.PSInvokeMemberBinder.InvokeAdaptedMember(Object obj, String methodName, Object[] args) in C:\Users\1\Documents\GitHub\iSazo
nov\PowerShell\src\System.Management.Automation\engine\runtime\Binding\Binders.cs:line 7483
   at System.Dynamic.UpdateDelegates.UpdateAndExecute1[T0,TRet](CallSite site, T0 arg0)
   at System.Management.Automation.Interpreter.DynamicInstruction`2.Run(InterpretedFrame frame) in C:\Users\1\Documents\GitHub\iSazonov\PowerShell\src\System.Managemen
t.Automation\engine\interpreter\DynamicInstructions.Generated.cs:line 141
   at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame) in C:\Users\1\Documents\GitHub\iSazonov\PowerShell\src\Syste
m.Management.Automation\engine\interpreter\ControlFlowInstructions.cs:line 355
TargetObject          : ToString
CategoryInfo          : InvalidOperation: (ToString:String) [], RuntimeException
FullyQualifiedErrorId : MethodNotFound
InvocationInfo        :
    ScriptLineNumber : 1
    OffsetInLine     : 1
    HistoryId        : 4
    Line             : $des.ToString()
    PositionMessage  : At line:1 char:1
                       + $des.ToString()
                       + ~~~~~~~~~~~~~~~
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
```

**Comment by iSazonov** [comment]
More simple repro:
```powershell
$a = [psobject]::new()
$a.<Ctrl-Tab>
$a.psobject.<Ctrl-Tab>

File: src/System.Management.Automation/engine/MshObjectIConvertible.cs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+        /// <inheritdoc/>
+        public object ToType(Type conversionType, IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToType(conversionType, provider);
+
+        private static IConvertible GetIConvertibleOrThrow(PSObject pso)
+        {
+            object obj = PSObject.Base(pso);
+            if (obj is PSObject || obj is not IConvertible value)
+            {
+                ThrowInvalidCastException();
```
Comment (SteveL-MSFT) [inline src/System.Management.Automation/engine/MshObjectIConvertible.cs:89]: Since this method is only used once here, would it be better to simply have this block throw the exception and not have the workaround?

**Review by daxian-dbw — COMMENTED** [review]
I did some debugging but couldn't find the root cause. It cannot resolve the `ToString` member from the deserialized PSObject instance. The `InstanceMethods` member of the PSObject instance also contains different entries comparing with when running in a build without this change.

Specifically, when running `$des.ToString`, the `_hasInstanceMember` is `false` with the changes in this PR while it's `true` when running without.
https://github.com/PowerShell/PowerShell/blob/6dfba03aa3d76590d28619d41631f2115899af78/src/System.Management.Automation/engine/runtime/Binding/Binders.cs#L5209-L5212

So, something very confusing may happen in the deserialization/serialization code.

Also, I found you can tab complete the `IConvertible` interface methods on `$des`, which is a `Deserialized.xxx` PSObject, but not on a newly created PSObject `[psobject]::new()`. But `$des.ToBoolean()` for example will fail too.

File: src/System.Management.Automation/engine/MshObjectIConvertible.cs [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        /// </remarks>
+        /// <returns>Returns <see cref="System.TypeCode"/> of the PSObject.</returns>
+        public TypeCode GetTypeCode()
+        {
+            object obj = PSObject.Base(this);
+            return obj is PSObject ? TypeCode.Object : Convert.GetTypeCode(obj);
+        }
+
+        /// <inheritdoc/>
+        public bool ToBoolean(IFormatProvider? provider) => GetIConvertibleOrThrow(this).ToBoolean(provider);
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/MshObjectIConvertible.cs:37]: All members should be explicit definitions, like `bool IConvertible.ToBoolean(IFormatProvider? provider) => ...`

However, changing to explicit definitions does address the exception we saw. You get the same error.

File: src/System.Management.Automation/engine/MshObjectIConvertible.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/MshObjectIConvertible.cs:89]: `throw` blocks inlining. .Net suggests long term patterns like ArgumentNullException.ThrowIf(). We can benefit from this too.

**Comment by iSazonov** [comment]
> I did some debugging but couldn't find the root cause.

Eh, Debugger initializes PowerShell caches and hides real steps. As result we don't see details. :-(

I added Console.WriteLine in the new code and I did not see they are called.

Then I searched for IConvertible and it only exists in PSObject.ToString(). This is not what we need.

I searched for IConvertible in .Net and there is nothing in the C# code either.

So I assumed that the reason is in the _.Net reflection code_, somewhere in C++.
In other words I assume that PowerShell calls type.GetMembers() somewhere to get the list of members of the type and the result depends on IConvertible. But I don't know where that point is in PowerShell. I hope you know. Perhaps the easiest way to find it is to tab-completion code.

**Comment by daxian-dbw** [comment]
> PowerShell calls type.GetMembers() somewhere

You can find that code in the adaptor code, `DotnetAdaptor` in particular for reflection operations.

**Comment by daxian-dbw** [comment]
The fact that PSObject representing a deserialized .NET object behaves differently from a regular PSObject is very bizarre.

**Comment by iSazonov** [comment]
1. Serialization fix
Serializer uses 
https://github.com/PowerShell/PowerShell/blob/66a89826bbaa7bb35c2938c3afc7785c469dbaba/src/System.Management.Automation/engine/serialization.cs#L5517-L5526
to get ToString method representation. It uses Convert.ToString() which uses IConvertible.ToString() on first place
https://github.com/dotnet/runtime/blob/967250cc86b4c613015066b47b882e7424fdb590/src/libraries/System.Private.CoreLib/src/System/Convert.cs#L1922-L1929
This is inside the .Net. This explains why we didn't see this problem in the debugger.
So I had to use PSObject.ToString() method as IConvertible.ToString() implementation.
I think it is more universal and more save than place the fix in GetToString().

3. Type conversion fix
The problem was in `PSObject -> type` cast. PowerShell uses compex type converters and one branch involves IConvertible https://github.com/PowerShell/PowerShell/blob/66a89826bbaa7bb35c2938c3afc7785c469dbaba/src/System.Management.Automation/engine/LanguagePrimitives.cs#L5674

Fix is to block the branch if fromType is PSObject so that return previous behavior.

Alternatively, we could change the IConvertible.ToType() implementation, but through the .Net code this would again take us back to the complex PowerShell type conversion code. This seems unpredictable, dangerous (recursion), and most importantly, unreasonable.

4. Thoughts about other IConvertible methods

The current implementation calls a method of base type (e.g., ToInt32()) for which the TypeCode is known beforehand. We could call the universal LanguagePrimitives.ConvertTo() method, but we cannot implement a dynamic value for TypeCode because of a contradiction - TypeCode must be known before the conversion, but we can only compute it after the conversion. This also contradicts the documentation requirement to return TypeCode.Object for all other types.

**Comment by daxian-dbw** [comment]
Thinking about this again, I'm not sure if making `PSObject` implement `IConvertible` is the right solution. There are many popularly used interfaces out there, so if we implement `IConvertible` to facilitate the cases where `PSObject` is wrapping an object that implements `IConvertible`, are we going to implement other interfaces to facilitate other cases?

**Comment by iSazonov** [comment]
We started with a specific user case, namely System.Data makes extensive use of IConvertible. This interface is an old legacy, it will never evolve. .Net team says to use more modern APIs.  I also found that ASP.NET also uses IConvertible extensively. 
So the big question is whether we want to support System.Data and ASP.NET natively (I don't know anything else). If yes, then it's PR worthwhile. If not, it should be closed.

**Comment by daxian-dbw** [comment]
Marked the PR "Engine - WG". Will let the engine WG review.

**Comment by iSazonov** [comment]
For additional information about IConvertible see https://github.com/dotnet/runtime/issues/73074#issuecomment-1200049733 and below comments.

**Comment by SeeminglyScience** [comment]
The Engine WG discussed this and decided that we should not implement interfaces that may or may not be implemented on the base object.

A primary concern is that a `is IConvertible` check will return true but a cast will throw. There are also a lot of other interfaces that would need to be implemented if we wanted to go down that path.

Instead we recommend that the original issue is fixed with either:

1. An implementation of `IDynamicInterfaceCastable` (which may or may not actually be feasible)
2. An explicit unwrapping of `PSObject` in the binder. Possibly as a special case for `DataTable`, but ideally in a way that matches our current handling when a `PSObject` wrapped object is an argument in a method.
